pkgs/development/tools/haskell/haskell-language-server/withWrapper.nix

```diff
@@ -13,9 +78,44 @@
 # for example. Read more about this in the haskell-language-server section of the nixpkgs manual.
 #
 let
-  inherit (lib) concatStringsSep concatMapStringsSep take splitString pipe optionals;
-  inherit (haskell.lib.compose) justStaticExecutables overrideCabal enableCabalFlag disableCabalFlag;
+  inherit (haskell.lib.compose)
+    justStaticExecutables
+    overrideCabal
+    enableCabalFlag
+    disableCabalFlag
+    ;
+
   getPackages = version: haskell.packages."ghc${version}";
+
+  # Given the list of `supportedFormatters`, remove every formatter that we know of (knownFormatters)
+  # by disabling the cabal flag and also removing the formatter libraries.
+  removeUnnecessaryFormatters =
+    let
+      # only formatters that were not requested
+      unwanted = lib.pipe knownFormatters [
+        (lib.filterAttrs (fmt: _: ! (lib.elem fmt supportedFormatters)))
+        lib.attrsToList
+      ];
+      # all flags to disable
+      flags = map (fmt: fmt.value.cabalFlag) unwanted;
+      # all dependencies to remove from hls
+      deps = lib.concatMap (fmt: fmt.value.packages) unwanted;
+
+      # null out elements where the predicate is false
+      nullElems = pred: map (v: if pred v then null else v);
+      # remove nulls from a list
+      stripNulls = lib.filter (x: x != null);
+
+      # remove all unwanted dependencies of formatters we don’t want
+      stripDeps = overrideCabal (drv: {
+        libraryHaskellDepends = lib.pipe (drv.libraryHaskellDepends or []) [
+          stripNulls
+          (nullElems (dep: lib.elem dep.pname deps))
```

Comment:
Aren't these upside down? You want to first mark the relevant elements as null, and then remove the nulls right?

I would maybe suggest using foldl for this kind of operation, but I'm unsure if that would cause a bunch more allocations than this technique, or if they are equally heavy.

Reply:
No, the existing list already has nulls, which I need to strip out first. Granted, I create more nulls afterwards, I should probably just filter instead.

Reply:
did that